clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy

```diff
@@ -16,35 +16,56 @@
 
 package com.netflix.spinnaker.clouddriver.openstack.client
 
+import com.netflix.spinnaker.clouddriver.openstack.deploy.exception.OpenstackOperationException
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperations
 import org.openstack4j.api.OSClient
+import org.openstack4j.model.common.ActionResponse
 
 /**
  * Provides access to the Openstack API.
+ *
+ * TODO tokens will need to be regenerated if they are expired.
```

Comment:
From a quick glance at documentation, 24 hours, though I need to verify that between v2 (uses access object) and v3 (uses token). Though I do think there is an expiration date on the token or access object that can be checked.

I would like to move the logic that connects and gets the token out of the factory and into `OpenstackClientProvider.getClient`. Then instead of a simple getter, it automatically handles reauthenticating for a new token if needed.

All this is to deal with using the `OSClient` object in a multi-threaded environment. Openstack4j seems to tie that object to a thread (well, throws errors if it is shared across threads), so we have to create a new client object for the openstack session using the cached token.